Source/CCZ4/NewConstraints.impl.hpp

```diff
@@ -0,0 +1,147 @@
+          template <typename> class diff2_vars_t>
+Constraints::Vars<data_t> Constraints::constraint_equations(
+    const vars_t<data_t> &vars, const vars_t<Tensor<1, data_t>> &d1,
+    const diff2_vars_t<Tensor<2, data_t>> &d2) const
+{
+    Vars<data_t> out;
+
+    auto h_UU = TensorAlgebra::compute_inverse_sym(vars.h);
+    auto chris = TensorAlgebra::compute_christoffel(d1.h, h_UU);
+
+    if (m_c_Ham >= 0 || m_c_Ham_abs_terms >= 0)
+    {
+        auto ricci = CCZ4Geometry::compute_ricci(vars, d1, d2, h_UU, chris);
+
+        auto A_UU = TensorAlgebra::raise_all(vars.A, h_UU);
+        data_t tr_A2 = TensorAlgebra::compute_trace(vars.A, A_UU);
+
+        out.Ham = ricci.scalar +
+                  (GR_SPACEDIM - 1.) * vars.K * vars.K / GR_SPACEDIM - tr_A2;
+        out.Ham -= 2 * m_cosmological_constant;
+
+        out.Ham_abs_terms =
+            abs(ricci.scalar) + abs(tr_A2) +
+            abs((GR_SPACEDIM - 1.) * vars.K * vars.K / GR_SPACEDIM);
+        out.Ham_abs_terms += 2.0 * abs(m_cosmological_constant);
+    }
+
+    if (m_c_Moms.size() > 0 || m_c_Moms_abs_terms.size() > 0)
+    {
+        Tensor<2, data_t> covd_A[CH_SPACEDIM];
+        FOR3(i, j, k)
+        {
+            covd_A[i][j][k] = d1.A[j][k][i];
+            FOR1(l)
+            {
+                covd_A[i][j][k] += -chris.ULL[l][i][j] * vars.A[l][k] -
+                                   chris.ULL[l][i][k] * vars.A[l][j];
+            }
+        }
+        FOR1(i)
+        {
+            out.Mom[i] = -(GR_SPACEDIM - 1.) * d1.K[i] / GR_SPACEDIM;
+            out.Mom_abs_terms[i] = abs(out.Mom[i]);
+        }
+        Tensor<1, data_t> covd_A_term = 0.0;
+        Tensor<1, data_t> d1_chi_term = 0.0;
+        const data_t chi_regularised = simd_max(1e-6, vars.chi);
+        FOR3(i, j, k)
+        {
+            covd_A_term[i] += h_UU[j][k] * covd_A[k][j][i];
+            d1_chi_term[i] += -GR_SPACEDIM * h_UU[j][k] * vars.A[i][j] *
+                              d1.chi[k] / (2 * chi_regularised);
+        }
+        FOR1(i)
+        {
+            out.Mom[i] += covd_A_term[i] + d1_chi_term[i];
+            out.Mom_abs_terms[i] += abs(covd_A_term[i]) + abs(d1_chi_term[i]);
+        }
+    }
+    return out;
+}
+
+template <class data_t>
+void Constraints::store_vars(Vars<data_t> &out,
+                             Cell<data_t> &current_cell) const
+{
+    if (m_c_Ham >= 0)
+        current_cell.store_vars(out.Ham, m_c_Ham);
+    if (m_c_Ham_abs_terms >= 0)
+        current_cell.store_vars(out.Ham_abs_terms, m_c_Ham_abs_terms);
+    if (m_c_Moms.size() == GR_SPACEDIM)
+    {
+        FOR1(i)
+        {
+            int ivar = m_c_Moms.begin() + i;
+            current_cell.store_vars(out.Mom[i], ivar);
+        }
+    }
+    else if (m_c_Moms.size() == 1)
+    {
+        data_t Mom_sq = 0.0;
+        FOR1(i) { Mom_sq += out.Mom[i] * out.Mom[i]; }
+        data_t Mom = sqrt(Mom_sq);
+        current_cell.store_vars(Mom, m_c_Moms.begin());
+    }
```

Comment:
Beautiful how you did this!

Reply:
Thanks but I wish there was a way to do this using templates to avoid branches in the `compute` function.